File: SOR/SOR.go
package main

import (
	"encoding/csv"
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

type Array []float64

func main() {
	args := os.Args
	for i := 1; i < len(args); i++ {
		file_name := args[i]
		fmt.Println(args[i], ":")
		sor_solver(readCSVFile(file_name)) //n = system dimensions, A = augmented matrix
	}
}
func readCSVFile(file_name string) ([]Array, []float64) {
	csvFile, err := os.Open(file_name)
	if err != nil {
		fmt.Println(err)
	}
	// fmt.Println("Successfully Opened CSV file")
	defer csvFile.Close()

	csvLines, err := csv.NewReader(csvFile).ReadAll()
	if err != nil {
		fmt.Println(err)
	}

	matrix := []Array{}
	b := []float64{}
	for _, line := range csvLines {
		line_length := len(line)
		array_aux := []float64{}
		for i := 0; i < line_length; i++ {
			line_i, err := strconv.ParseFloat(line[i], 64)
			if err != nil {
				fmt.Println(err)
			}
			if i == line_length-1 {
				b = append(b, line_i)
				break
			}
			array_aux = append(array_aux, line_i)
		}
		matrix = append(matrix, array_aux)
	}

	return matrix, b
}

func euclidian_distance(A []float64, b []float64) float64 {
	var distance = 0.0
	for fila := 0; fila < len(A); fila++ {
		distance = distance + math.Pow(A[fila]-b[fila], 2)
	}

	distance = math.Sqrt(distance)
	return distance
}
func dot(A []Array, b []float64, l int) []float64 {

	result := make([]float64, l)
	for i := 0; i < l; i++ {
		var aux = 0.0
		for j := 0; j < l; j++ {
			aux = aux + b[j]*A[i][j]
		}
		result[i] = aux
	}
	return result
}
func substraction(a []float64, b []float64, l int) []float64 {

	result := make([]float64, l)
	for i := 0; i < l; i++ {
		result[i] = a[i] - b[i]
	}
	return result
}

func sor_solver(A []Array, b []float64) {

	//Create where the solution will be store
	var long int = len(b)
	var X0 = make([]float64, long)

	var tolerancia float64 = 0.000001 //1.0071743059050822
	var iteracionMax int = 100

	var n int = len(A)
	var m int = len(A)
	var omega float64 = 1.5
	var X = X0
	var tol float64 = euclidian_distance(dot(A, X, n), b)

	var iteracion int = 0
	start := time.Now()

	for tol > tolerancia && iteracion < iteracionMax {

		for fila := 0; fila < n; fila++ {
			var suma float64 = 0
			for columna := 0; columna < m; columna++ {
				if fila != columna {
					suma = (A[fila][columna]) * (X[columna])
				}
			}
			var nuevo float64 = (1.0-omega)*X[fila] + (omega/A[fila][fila])*(b[fila]-suma)

			X[fila] = nuevo

		}

		tol = euclidian_distance(dot(A, X, n), b)
		iteracion = iteracion + 1
	}
	// fmt.Println("Respuesta X: ", X)
	fmt.Println("")
	elapsed := time.Since(start)
	fmt.Println(elapsed.Nanoseconds())
	// fmt.Println("EXECUTION TIME: ", elapsed.Nanoseconds(), " nanoseconds")
}

File: SOR/SOR.scala
import scala.io.Source
import javax.print.DocFlavor.INPUT_STREAM
import javax.print.attribute.standard.PrinterURI
import math._
import java.io.File
object SOR {
	def main(args: Array[String]): Unit = {
        for(arg<-args){
            var param = readCSVFile(arg)
            var (aug_matrix) = param
            sor_solver(aug_matrix)
        }
    }
    def readCSVFile(file: String): (Array[Array[Double]]) = {
        var C = io.Source.fromFile(file).getLines().map(_.split(",").map(_.trim.toDouble)).toArray
        var n = C.length
        var m = C(0).length
        var aug_matrix = Array.ofDim[Double](n, m)

        for(i <- 0 to n-1){
           for (j <- 0 to n){
              aug_matrix(i)(j) = C(i)(j)
           }
        }
        return (aug_matrix)
    }
    def sor_solver(aug_matrix: Array[Array[Double]]): Unit = {
        
        var B = Array.ofDim[Double](aug_matrix.length,1)
        var BtoArray = Array.fill(aug_matrix.length){0.0}
       
          for(i <- 0 to B.length-1){
                  B(i)(0) = aug_matrix(i)(aug_matrix(1).length-1)
                  BtoArray.update(i,B(i)(0))
          }
        var A = Array.ofDim[Double](aug_matrix.length,aug_matrix(1).length-1)

        for(i <- 0 to aug_matrix.length-1){
             for(j <- 0 to aug_matrix(1).length-2){
                A(i)(j) = aug_matrix(i)(j)
             }
          }
          for(i <- 0 to aug_matrix.length-1){
             for(j <- 0 to aug_matrix(1).length-2){
                A(i)(j) = aug_matrix(i)(j)
             }
          }

	    var tol = Array.ofDim[Double](aug_matrix.length,aug_matrix.length)
        var X0  = Array.ofDim[Double](aug_matrix.length)
        
        var matMul = A.zip(tol) map (_.zipped map (_ * _)) map (_.sum)
        
        var aiuda = distance(matMul,BtoArray)

        var tolerancia: Double = 0.000001
        var iteracionMax: Int = 100
        var Omega: Double = 0.5
        
        var n: Int = aug_matrix.size    
        var m: Int = aug_matrix.size

        var X = X0
        var iteracion : Int = 0
        
        var time_ini = System.nanoTime()

        while ((aiuda > tolerancia || iteracion > 100)){
            for(fila <- 0 to n - 1){
                var suma: Double = 0
                var sigma: Double = 0
                for(columna <- 0 to m - 1){
                    if(fila != columna){
                        sigma = sigma + ((A(fila)(columna)) * (X(columna))) 
                    }
                }
                var phi = (1-Omega) * X(fila) + (Omega / A(fila)(fila)) * (B(fila)(0)-sigma)

                X.update(fila, phi)                
                tol.update(fila,X)
            }
            var matMul = A.zip(tol) map (_.zipped map (_ * _)) map (_.sum)

            aiuda = distance(matMul,BtoArray)
            iteracion = iteracion + 1
        }     
        
        var Xt = Array(Array(X(0)))
        for(i <- 1 to X.length-1){
            Xt ++= Array(Array(X(i)))
        }
        
        var duration = (System.nanoTime - time_ini) / 1e9d

        println("EXECUTION TIME:" + duration + " nanoseconds")

     }

    def distance(xs: Array[Double], ys: Array[Double]) = {
        sqrt((xs zip ys).map { case (x,y) => pow(y - x, 2) }.sum)
    }   

}
